chips/nrf51/src/radio.rs

```diff
@@ -0,0 +1,456 @@
+            // this state only verifies that END is received in TX-mode
+            // which means that the transmission is finished
+            match regs.STATE.get() {
+                RADIO_STATE_TXRU |
+                RADIO_STATE_TXIDLE |
+                RADIO_STATE_TXDISABLE |
+                RADIO_STATE_TX => {
+                    match regs.FREQEUNCY.get() {
+                        // frequency 39
+                        80 => {
+                            self.radio_off();
+                        }
+                        // frequency 38
+                        26 => {
+                            self.set_channel_freq(39);
+                            self.set_data_white_iv(39);
+                            regs.READY.set(0);
+                            regs.TXEN.set(1);
+                        }
+                        // frequency 37
+                        2 => {
+                            self.set_channel_freq(38);
+                            self.set_data_white_iv(38);
+                            regs.READY.set(0);
+                            regs.TXEN.set(1);
+                        }
+                        // don't care as we only support advertisements at the moment
+                        _ => {
+                            self.set_channel_freq(37);
+                            self.set_data_white_iv(37)
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        self.enable_nvic();
+        self.enable_interrupts();
+    }
+
+
+    pub fn enable_interrupts(&self) {
+        // INTENSET REG
+        let regs = unsafe { &*self.regs };
+        // 15 i.e set 4 LSB
+        regs.INTENSET.set(1 | 1 << 1 | 1 << 2 | 1 << 3);
+    }
+
+    pub fn disable_interrupts(&self) {
+        let regs = unsafe { &*self.regs };
+        // disable all possible interrupts
+        regs.INTENCLR.set(0x4ff);
+    }
+
+    pub fn enable_nvic(&self) {
+        nvic::enable(NvicIdx::RADIO);
+    }
+
+    pub fn disable_nvic(&self) {
+        nvic::disable(NvicIdx::RADIO);
+    }
+
+    pub fn reset_payload(&self) {
+        // reset contents except header || address
+        for i in 9..39 {
+            unsafe {
+                PAYLOAD[i] = 0;
+            }
+        }
+        unsafe {
+            PAYLOAD[1] = 6;
+        }
+    }
+
+    // these are used by ble_advertising_driver and are therefore public
+    // FIXME: should be moved to the HIL later
+
+    pub fn start_adv(&self) {
+        self.init_radio_ble();
+    }
+
+    pub fn set_adv_data(&self,
+                        ad_type: usize,
+                        data: &'static mut [u8],
+                        len: usize,
+                        offset: usize)
+                        -> &'static mut [u8] {
+        if offset == 9 {
+            //FIXME: move call to the capsule!?
+            self.reset_payload();
+        }
+        unsafe {
+            PAYLOAD[offset] = (len + 1) as u8;
```

Comment:
Also, FWIW, rust will insert a bounds check anyway, and if you do the check explicitly (in such a way that LLVM can notice) the bounds check will be optimized out, so there's not really a "cost" in most cases to adding an explicit check.